File: bootstrap/src/types/collections/list.rs
/* Serene --- Yet an other Lisp
*
* Copyright (c) 2020  Sameer Rahmani <lxsameer@gnu.org>
*
* Exprhis program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 2 of the License.
*
* Exprhis program is distributed in the hope that it will be useful,
* but WIExprHOUExpr ANY WARRANExprY; without even the implied warranty of
* MERCHANExprABILIExprY or FIExprNESS FOR A PARExprICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
//use crate::builtins::def;
use crate::ast::{Expr, Expression, PossibleExpr};
use crate::errors::err;
use crate::runtime::RT;
use crate::scope::Scope;
use crate::types::collections::core::Seq;
use crate::types::Symbol;

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct List {
    elements: Vec<Expr>,
}

impl List {
    pub fn new_empty() -> List {
        List { elements: vec![] }
    }

    pub fn new(elems: &[Expr]) -> List {
        List {
            elements: elems.to_vec(),
        }
    }

    pub fn push(&mut self, elem: Expr) {
        self.elements.push(elem)
    }

    pub fn count(&self) -> usize {
        self.elements.len()
    }
}

impl Seq<Expr> for List {
    type Coll = Self;

    fn first(&self) -> Option<Expr> {
        match self.elements.first() {
            Some(e) => Some(e.clone()),
            None => None,
        }
    }

    fn rest(&self) -> List {
        if self.count() > 0 {
            List::new(&self.elements[1..])
        } else {
            List::new_empty()
        }
    }
}

impl Expression for List {
    fn eval(&self, rt: &RT, scope: &Scope) -> PossibleExpr {
        if self.count() == 0 {
            return Ok(Expr::Nil);
        }

        let first = self.first().unwrap();
        let rest = self.rest();
        Err(err("NotImplemented".to_string()))
        //Ok(Expr::Cons(Box::new(*self)))
        // match first {
        //     Expr::Sym(sum) => {}
        //     _ => Err(err("NotImplemented".to_string())),
        // }
    }
}

File: src/libserene/exprs/list.cpp
/*
 * Serene Programming Language
 *
 * Copyright (c) 2019-2022 Sameer Rahmani <lxsameer@gnu.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "serene/exprs/list.h"

#include "serene/errors.h"
#include "serene/exprs/call.h"
#include "serene/exprs/def.h"
#include "serene/exprs/expression.h"
#include "serene/exprs/fn.h"
#include "serene/exprs/symbol.h"

#include <llvm/Support/Casting.h>
#include <llvm/Support/ErrorHandling.h>
#include <llvm/Support/FormatVariadic.h>

#include <iterator>
#include <utility>

namespace serene {
namespace exprs {

List::List(const List &l) : Expression(l.location) {
  this->elements = l.elements;
};

List::List(const reader::LocationRange &loc, Node &e) : Expression(loc) {
  elements.push_back(e);
};

List::List(const reader::LocationRange &loc, Ast elems)
    : Expression(loc), elements(std::move(elems)){};

ExprType List::getType() const { return ExprType::List; };

std::string List::toString() const {
  std::string s{this->elements.empty() ? "-" : ""};

  for (const auto &n : this->elements) {
    s = llvm::formatv("{0} {1}", s, n->toString());
  }

  return llvm::formatv("<List {0}>", s);
};

MaybeNode List::analyze(semantics::AnalysisState &state) {

  if (!elements.empty()) {
    auto *first = elements[0].get();

    if (first->getType() == ExprType::Symbol) {
      auto *sym = llvm::dyn_cast<Symbol>(first);

      if (sym != nullptr) {
        if (sym->name == "def") {
          return Def::make(state, this);
        }

        if (sym->name == "fn") {
          return Fn::make(state, this);
        }
      }
    }

    return Call::make(state, this);
  }

  return EmptyNode;
};

bool List::classof(const Expression *e) {
  return e->getType() == ExprType::List;
};

std::vector<Node>::const_iterator List::cbegin() { return elements.begin(); }

std::vector<Node>::const_iterator List::cend() { return elements.end(); }

std::vector<Node>::iterator List::begin() { return elements.begin(); }

std::vector<Node>::iterator List::end() { return elements.end(); }

size_t List::count() const { return elements.size(); }

llvm::Optional<Expression *> List::at(uint index) {
  if (index >= elements.size()) {
    return llvm::None;
  }

  return llvm::Optional<Expression *>(this->elements[index].get());
}

Ast List::from(uint index) {
  if (index < elements.size()) {
    return Ast(elements.begin() + index, elements.end());
  }

  return Ast();
}

void List::append(Node n) { elements.push_back(std::move(n)); }
} // namespace exprs
} // namespace serene
